### PR TITLE
Feature refactor saved movies and movie info

### PR DESCRIPTION
**Pull request by smabhena**
Refactored Movie Info and Saved Movie features
- Created a single repository that is shared 
- Refactored Movie Info feature
- Refactored Saved Movie feature
- Fixed delete movie bug

### CODE REVIEW

File: Mooov/Mooov/Save Movie/SaveMovieViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                self.delegate?.reloadView()
+            }
+            
+        } catch {
+            self.delegate?.show(error: "Try again")
+        }
+    }
+    
+    func deleteMovieItem(_ movieItem: MovieItem, _ index: Int) {
+        guard let context = Constants.viewContext else {
```
Comment (bgouws) [inline Mooov/Mooov/Save Movie/SaveMovieViewModel.swift:48]: You can move code like this to a file called `SaveMovieRepository`

**Review by bgouws — CHANGES_REQUESTED** [review]
Some clean up 😄

File: Mooov/Mooov/Movie Info/MovieInfoViewController.swift [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+class MovieInfoViewController: UIViewController {
+    @IBOutlet private weak var imageView: UIImageView!
+    @IBOutlet private weak var titleLabel: UILabel!
+    @IBOutlet private weak var genreLabel: UILabel!
+    @IBOutlet private weak var runtimeLabel: UILabel!
+    @IBOutlet private weak var ratingLabel: UILabel!
+    @IBOutlet private weak var plotLabel: UITextView!
+    @IBOutlet private weak var savedMovieButton: UIButton!
+    
+    private var id: String?
```
Comment (bgouws) [inline Mooov/Mooov/Movie Info/MovieInfoViewController.swift:19]: Do we need to have these variables here?

File: Mooov/Mooov/Movie Info/MovieInfoViewController.swift [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        if let movieTitle = movieTitle, let image = image {
+            viewModel.isMovieSaved(movieTitle, image)
+        }
+    }
+    
+    @IBAction private func tappedAddButton() {
+        viewModel.createMovieItem(movieTitle, image)
+    }
+    
+    func setMovieId(_ id: String) {
```
Comment (bgouws) [inline Mooov/Mooov/Movie Info/MovieInfoViewController.swift:44]: Rather pass this through as one object and then set it in your `ViewModel`. 😄

File: Mooov/Mooov/Movie Info/MovieInfoViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        self.delegate = delegate
+        self.repository = repository
+    }
+    
+    var movieData: MovieInfo? {
+        return data
+    }
+    
+    func createMovieItem(_ title: String?, _ image: String?) {
+        guard let title = title, let image = image, let context = Constants.viewContext else { return }
```
Comment (bgouws) [inline Mooov/Mooov/Movie Info/MovieInfoViewModel.swift:31]: Why not call the repository here? 😄

File: Mooov/Mooov/Movie Info/MovieInfoViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            try context.save()
+            self.delegate?.disableButton()
+        } catch {
+            self.delegate?.showError(error: "Try again")
+        }
+    }
+    
+    func isMovieSaved(_ movieTitle: String, _ movieImage: String) {
+        do {
+            guard let movies = try Constants.viewContext?.fetch(MovieItem.fetchRequest()) else { return }
```
Comment (bgouws) [inline Mooov/Mooov/Movie Info/MovieInfoViewModel.swift:47]: Same as above here 😄

File: Mooov/Mooov/Save Movie/SaveMovieViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    var movieList: [MovieItem]? {
+        return movies
+    }
+    
+    func movie(atIndex: Int) -> MovieItem? {
+        return movies?[atIndex] 
+    }
+    
+    func allSavedMovies() {
+        repository?.fetchSavedMovies(completion: { [weak self] movies in
```
Comment (bgouws) [inline Mooov/Mooov/Save Movie/SaveMovieViewModel.swift:38]: Very nice! 😄

**Review by bgouws — CHANGES_REQUESTED** [review]
Looking good, maybe just update that one `ViewController`  to use the repo rather than access `CoreData` directly.

File: Mooov/Mooov/Movie Info/MovieInfoViewController.swift (hunk 2 above)
Comment (smabhena) [reply Mooov/Mooov/Movie Info/MovieInfoViewController.swift:19]: Yeah they're used to hold the data sent from the SearchResultsViewController

File: Mooov/Mooov/Movie Info/MovieInfoViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+protocol MovieInfoViewModelDelegate: AnyObject {
+    func showError(error: String)
+    func disableButton()
+    func loadContent()
+}
+
+class MovieInfoViewModel {
+    private weak var delegate: MovieInfoViewModelDelegate?
+    private var movieRepository: MovieRepositoryType?
+    private var coreDataRepository: CoreDataRepository?
```
Comment (bgouws) [inline Mooov/Mooov/Movie Info/MovieInfoViewModel.swift:19]: Use a `protocol` to abstract this `repository` 😄

File: Mooov/Mooov/Movie Info/MovieInfoViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    private var data: MovieInfo?
+    private var movie: Movie?
+    
+    init(movieRepository: MovieRepository,
+         coreDataRepository: CoreDataRepository,
```
Comment (bgouws) [inline Mooov/Mooov/Movie Info/MovieInfoViewModel.swift:24]: Here too 😄

File: Mooov/Mooov/Movie Info/MovieInfoViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        return data
+    }
+    
+    func setMovie(_ movie: Movie) {
+        self.movie = movie
+    }
+    
+    func createMovie() {
+        guard let movie = movie else { return }
+        coreDataRepository?.createMovieItem(movie: movie, completion: { [weak self] result in
```
Comment (bgouws) [inline Mooov/Mooov/Movie Info/MovieInfoViewModel.swift:41]: Nice!

File: Mooov/Mooov/Save Movie/SaveMovieViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+import Foundation
+
+protocol SaveMovieViewDelegate: AnyObject {
+    func reloadView()
+    func show(error: String)
+}
+
+class SaveMovieViewModel {
+    private weak var delegate: SaveMovieViewDelegate?
+    private var repository: CoreDataRepository?
```
Comment (bgouws) [inline Mooov/Mooov/Save Movie/SaveMovieViewModel.swift:17]: You can also abstract the type here with a `protocol`

File: Mooov/Mooov/Save Movie/SaveMovieViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    private var movies: [MovieItem]? = []
+    
+    init(delegate: SaveMovieViewDelegate, repository: CoreDataRepository) {
```
Comment (bgouws) [inline Mooov/Mooov/Save Movie/SaveMovieViewModel.swift:20]: Here too 😄 and parameters on their own line 👍

**Review by bgouws — CHANGES_REQUESTED** [review]
Some more clean up

File: Mooov/Mooov/Save Movie/SaveMovieViewModel.swift (hunk 1 above)
Comment (kirsty14) [reply Mooov/Mooov/Save Movie/SaveMovieViewModel.swift]: I agree with this 👍🏻

File: Mooov/Mooov/Save Movie/SaveMovieViewModel.swift (hunk 10 above)
Comment (kirsty14) [reply Mooov/Mooov/Save Movie/SaveMovieViewModel.swift:17]: I agree with this 👍🏻

File: Mooov/Mooov/Movie Info/MovieInfoViewModel.swift (hunk 8 above)
Comment (kirsty14) [reply Mooov/Mooov/Movie Info/MovieInfoViewModel.swift:24]: I agree with this 👍🏻

File: Mooov/Mooov/Movie Info/MovieInfoViewModel.swift (hunk 7 above)
Comment (kirsty14) [reply Mooov/Mooov/Movie Info/MovieInfoViewModel.swift:19]: I agree with this 👍🏻

File: Mooov/Mooov/Save Movie/SaveMovieViewModel.swift (hunk 11 above)
Comment (kirsty14) [reply Mooov/Mooov/Save Movie/SaveMovieViewModel.swift:20]: I agree with this 👍🏻

**Review by kirsty14 — APPROVED** [review]
Looking good! 👍🏻

**Review by bgouws — APPROVED** [review]

**Pull request merged**